exec/java-exec/src/main/java/org/apache/drill/exec/work/filter/RuntimeFilterManager.java

```diff
@@ -0,0 +1,755 @@
+
+    private boolean contain = false;
+
+    public WrapperOperatorsVisitor(PhysicalOperator targetOp, Fragment fragment) {
+      this.targetOp = targetOp;
+      this.fragment = fragment;
+    }
+
+    @Override
+    public Void visitExchange(Exchange exchange, Void value) throws RuntimeException {
+      List<Fragment.ExchangeFragmentPair> exchangeFragmentPairs = fragment.getReceivingExchangePairs();
+      for (Fragment.ExchangeFragmentPair exchangeFragmentPair : exchangeFragmentPairs) {
+        boolean same = exchange == exchangeFragmentPair.getExchange();
+        if (same) {
+          return null;
+        }
+      }
+      return exchange.getChild().accept(this, value);
+    }
+
+    @Override
+    public Void visitOp(PhysicalOperator op, Void value) throws RuntimeException {
+      boolean same = op == targetOp;
+      if (!same) {
+        for (PhysicalOperator child : op) {
+          child.accept(this, value);
+        }
+      } else {
+        contain = true;
+      }
+      return null;
+    }
+
+    public boolean isContain() {
+      return contain;
+    }
+  }
+
+  private boolean containsPhysicalOperator(Wrapper wrapper, PhysicalOperator op) {
+    WrapperOperatorsVisitor wrapperOpsVistitor = new WrapperOperatorsVisitor(op, wrapper.getNode());
+    wrapper.getNode().getRoot().accept(wrapperOpsVistitor, null);
+    return wrapperOpsVistitor.isContain();
+  }
+
+  private Wrapper findPhysicalOpContainer(Wrapper wrapper, PhysicalOperator op) {
+    boolean contain = containsPhysicalOperator(wrapper, op);
+    if (contain) {
+      return wrapper;
+    }
+    List<Wrapper> dependencies = wrapper.getFragmentDependencies();
+    if (CollectionUtils.isEmpty(dependencies)) {
+      return null;
+    }
+    for (Wrapper dependencyWrapper : dependencies) {
+      Wrapper opContainer = findPhysicalOpContainer(dependencyWrapper, op);
+      if (opContainer != null) {
+        return opContainer;
+      }
+    }
+    //should not be here
+    return null;
+  }
+
+  private static boolean isEqJoin(HashJoinPOP joinPOP) {
+    List<JoinCondition> joinConditions = joinPOP.getConditions();
+    for (JoinCondition joinCondition : joinConditions) {
+      String relationShip = joinCondition.getRelationship();
+      boolean equalRelationShip = "==".equals(relationShip) || "EQUALS".equalsIgnoreCase(relationShip);
+      if (!equalRelationShip) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private static class HashJoinOpHolder {
+
+    private int joinOpId;
+
+    private int joinMajorId;
+
+    private HashJoinPOP hashJoinPOP;
+
+    private int probeSideScanMajorId;
+
+    private int probeSideScanOpId;
+
+    private PhysicalOperator probeSideScanOp;
+
+    private List<CoordinationProtos.DrillbitEndpoint> probeSideScanEndpoints;
+
+    //probe side's children have blocked nodes.
+    private boolean meetBlockedOperator;
+
+    //whethe this join operator is a partitioned HashJoin or broadcast HashJoin
```

Comment:
In addition to broadcast and partitioned hash join, there's a third state also which is non-distributed hash join (this depends on slice_target, estimated row count etc.).  Is the JPPD not applied for such cases ?

Reply:
You mean there's only one HashJoin execution node? To that case , I guess JPPD is not necessary, since non-distributed one means less data, it's not cost efficient to generate a BF.

Reply:
Yeah, it would be fine to not do JPPD for that case .. we can document the expected behavior.